### PR TITLE
11 data scope funded research gtr

### LINKED ISSUE #11: Data scope (funded research): GtR
https://gtr.ukri.org/

### PR DESCRIPTION
**Pull request by ampudia19**
### Description

In this PR, I've included a major refactor of our data collection code for OpenAlex (OA), and introduced new modules for data collection and processing for Gateway to Research (GtR).

### Key Changes

1. **OA Data Collection Code Refactor:**
   - I've restructured the OA data collection code to enhance its modularity and functionality. This includes moving some parts to `utils.py`.

2. **Introduction of GtR Data Collection and Processing:**
   - New modules for fetching and processing GtR data have been added.

### Specific Requests

- **Jack's Review on OA Code:**
   - Jack, could you please check the functionality of the old OA code, especially the parts I've moved around? I've tried to make sure nothing breaks, but you never know. Adding some tests on your code may help in the future.
   - Additionally, there's a bit of redundancy in the code, particularly between the two similar request functions. It would be great if you could look into yours, see whether your code can use mine instead.

- **Harmonisation of Hierarchical Naming:**
   - I've noticed that the hierarchical naming in both parameters and catalog names could be better aligned. Currently, I've kept yours as top-level, but ideally, they should be nested under `<oa>`. See the other parameters for reference.
   - Any suggestions on restructuring these for consistency and clarity is appreciated :)

### CODE REVIEW

File: conf/base/catalog.yml
Comment (Jack-0-0) [inline conf/base/catalog.yml:1]: The new format `catalog.yml` looks good. I didn't know about yaml anchors and aliases.

File: src/alphafold_impact/pipelines/data_collection_gtr/pipeline.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    pipelines = []
+    for label in settings.DYNAMIC_PIPELINES_MAPPING["gtr"]:
+        pipelines.append(
+            pipeline(
+                template_pipeline,
+                parameters={"params:test": "test"},
+                namespace=f"gtr.data_collection.{label}",
+                tags=[label.rsplit('/', maxsplit=1)[-1], "gtr"],
+            )
+        )
```
Comment (Jack-0-0) [inline src/alphafold_impact/pipelines/data_collection_gtr/pipeline.py]: Can use a list comprehension here:
```suggestion
     pipelines = [
        pipeline(
            template_pipeline,
            parameters={"params:test": "test"},
            namespace=f"gtr.data_collection.{label}",
            tags=[label.rsplit('/', maxsplit=1)[-1], "gtr"],
        )
        for label in settings.DYNAMIC_PIPELINES_MAPPING["gtr"]
    ]
```

File: src/alphafold_impact/pipelines/data_collection_gtr/utils.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
+"""
+Utility functions for the GtR data collection pipeline.
+
+Functions:
+    _api_config: Constructs the API configuration dictionary for a given endpoint.
+    _extract_main_address: Extracts the main address from the addresses list.
+    _extract_value_from_nested_dict: Extracts a value from a nested dictionary based 
+        on the given keys and value.
+"""
```
Comment (Jack-0-0) [inline src/alphafold_impact/pipelines/data_collection_gtr/utils.py]: This docstring is missing a function description for `transform_nested_dict`.
Also, the functions in the docstring start with `_` but the functions do not.

File: src/alphafold_impact/pipelines/data_collection_gtr/nodes.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+            break
+
+        logger.info("Fetched page %s from %s", page, endpoint)
+        page += 1
+        time.sleep(random.randint(2, 3))  # [HACK] Respect web etiquette
+
+    return all_data
+
+
+def preprocess_data_to_df(raw_data: List[Dict[str, Any]], endpoint) -> pd.DataFrame:
```
Comment (Jack-0-0) [inline src/alphafold_impact/pipelines/data_collection_gtr/nodes.py]: `endpoint` is missing a typehint.

File: src/alphafold_impact/pipelines/data_collection_gtr/nodes.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+    """Preprocess data to a DataFrame.
+
+    Args:
+        raw_data (List[Dict[str, Any]]): The raw data in a list of dictionaries.
```
Comment (Jack-0-0) [inline src/alphafold_impact/pipelines/data_collection_gtr/nodes.py:221]: `endpoint` is missing from the docstring

File: src/alphafold_impact/pipelines/data_collection_gtr/nodes.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+                "lead_funder",
+                "lead_org_department",
+                "research_topics",
+                "research_subjects",
+            ]
+        ]
+
+
+def fetch_gtr_data(
+    parameters: Dict[str, Union[str, int]], endpoint, test: bool = False
```
Comment (Jack-0-0) [inline src/alphafold_impact/pipelines/data_collection_gtr/nodes.py:163]: `endpoint` is missing a typehint

File: src/alphafold_impact/pipelines/data_collection_gtr/nodes.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+) -> List[Dict[str, Any]]:
+    """Fetch data from the GtR API.
+
+    Args:
+        parameters (Dict[str, Union[str, int]]): Parameters for the API request.
+        endpoint (str): The endpoint to fetch data from.
```
Comment (Jack-0-0) [inline src/alphafold_impact/pipelines/data_collection_gtr/nodes.py:169]: `test` is missing from the docstring

File: src/alphafold_impact/pipelines/data_collection_oa/README.md
Comment (Jack-0-0) [inline src/alphafold_impact/pipelines/data_collection_oa/README.md:1]: This is nicer than before 🙂

File: src/alphafold_impact/pipelines/data_collection_oa/nodes.py [hunk 7]
```diff
@@ -312,11 +134,85 @@ def retrieve_oa_works_for_concepts_and_years(
...
+            for work in call:
+                ref_works = [
+                    w.replace("https://openalex.org/", "")
+                    for w in work["referenced_works"]
+                ]
+                work_id = work.get("id").replace("https://openalex.org/", "")
+                doi = work.get("doi").replace("https://doi.org/", "")
+                oa_doi_dict[work_id] = {}
+                oa_doi_dict[work_id]["doi"] = doi
+                oa_doi_dict[work_id]["referenced_works"] = ref_works
```
Comment (Jack-0-0) [inline src/alphafold_impact/pipelines/data_collection_oa/nodes.py:212]:
```suggestion
oa_doi_dict[work_id] = {"doi": doi, "referenced_works": ref_works}
```

File: src/alphafold_impact/pipelines/data_collection_oa/pipeline.py [hunk 8]
```diff
@@ -1,130 +1,166 @@
...
+        ]
+    )
 
-outgoing_pipeline = mpl(
-    pipe=template_pipeline,
-    parameters={"params:direction.template": "params:direction.outgoing"},
-    outputs={"template": "oa_int_works_incoming_citations"},
-    tags="outgoing",
-)
+    pipelines = []
```
Comment (Jack-0-0) [inline src/alphafold_impact/pipelines/data_collection_oa/pipeline.py]: Can use a list comprehension here

File: src/alphafold_impact/pipelines/data_processing_gtr/nodes.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
+"""
+This module contains the functions used in the data processing pipeline for
+    the Gateway to Research data.
+
+The functions are used to load the data, process it, and save it in the
+    required format.
+"""
```
Comment (Jack-0-0) [inline src/alphafold_impact/pipelines/data_processing_gtr/nodes.py]: This doesn't have the function descriptions like the other modules.

File: src/alphafold_impact/pipelines/data_processing_gtr/nodes.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+                        institution_list.append(triple)
+                    institutions_dict[work_id][
+                        author["author"]["id"].split("/")[-1]
+                    ] = institution_list
+                except KeyError:
+                    pass
+    return institutions_dict, publications_dict
+
+
+def process_institutions(gtr_authors: List[dict]):
```
Comment (Jack-0-0) [inline src/alphafold_impact/pipelines/data_processing_gtr/nodes.py]: Missing a return typehint here

File: src/alphafold_impact/pipelines/data_processing_gtr/nodes.py [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+            "author_id",
+            "institution_id",
+            "institution_name",
+            "country_code",
+        ],
+    )
+
+    return data_institutions
+
+def process_publications(gtr_publications: List[dict]):
```
Comment (Jack-0-0) [inline src/alphafold_impact/pipelines/data_processing_gtr/nodes.py]: Missing a return typehint here

File: src/alphafold_impact/pipelines/data_processing_gtr/nodes.py [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+            "abstract",
+            "doi",
+            "publication_date",
+        ],
+    )
+
+    return data_publications
+
+
+def process_alphafold_citations(reference_dict, cites_data):
```
Comment (Jack-0-0) [inline src/alphafold_impact/pipelines/data_processing_gtr/nodes.py]: Missing typehints here

**Review by Jack-0-0 — CHANGES_REQUESTED** [review]
@ampudia19 I have reviewed the PR and left some comments.

In a separate PR, I will make changes relating to my parts of the OA code so that they fit with the new conventions.

File: conf/base/catalog.yml
Comment (ampudia19) [reply conf/base/catalog.yml:1]: I didn't either! Quite handy when expected to keep all dataset references in a single file :)

**Pull request merged**